src/main/java/com/cloudbees/hudson/plugins/folder/AbstractFolder.java

```diff
@@ -337,189 +334,73 @@ public static <K, V extends TopLevelItem> Map<K, V> loadChildren(AbstractFolder<
         final ChildNameGenerator<AbstractFolder<V>,V> childNameGenerator = parent.childNameGenerator();
         Map<String,V> byDirName = new HashMap<>();
         if (parent.items != null) {
-            if (childNameGenerator == null) {
-                for (V item : parent.items.values()) {
-                    byDirName.put(item.getName(), item);
-                }
-            } else {
-                for (V item : parent.items.values()) {
-                    String itemName = childNameGenerator.dirNameFromItem(parent, item);
-                    if (itemName == null) {
-                        itemName = childNameGenerator.dirNameFromLegacy(parent, item.getName());
-                    }
-                    byDirName.put(itemName, item);
-                }
+            for (V item : parent.items.values()) {
+                byDirName.put(childNameGenerator.dirNameFromItem(parent, item), item);
```

Comment:
`dirNameFromItem` is declared to return null (though the reason is undocumented), and actual impls in `branch-api` do so at least in some (perhaps rare) conditions, so you still need to check for that I suppose. Maybe have a wrapper that uses `LegacyChildNameGenerator` as a fallback in case the real impl returns null from `itemNameFromItem` or `dirNameFromItem`?